# Financial Functions — Week 1
**Aayan Mulani | Decimal Point Analytics Preparation**

This notebook implements three core financial functions from scratch using Python. Building these manually (rather than using a library) develops an understanding of the underlying mathematics — essential for financial modelling roles.

**Functions covered:**
1. Compound Interest Table
2. EMI (Equated Monthly Instalment) Calculator
3. Moving Average

## 1. Compound Interest
**Simple interest** pays interest only on the original principal. **Compound interest** pays interest on the principal *and* on the accumulated interest — i.e. interest on interest.

**Formula:** A = P × (1 + r)ⁿ
- A = final amount
- P = principal (initial investment)
- r = annual interest rate (as a decimal)
- n = number of years

Compounding is why long-term investing is powerful. The interest earned in later years is far greater than in early years because the base keeps growing.

In [1]:
def compound_interest(principal, annual_rate, years):
    """
    Print a year-by-year compound interest table.
    principal   : initial investment amount (Rs)
    annual_rate : annual interest rate as decimal (e.g. 0.085 for 8.5%)
    years       : investment horizon in years
    """
    print(f'Principal: Rs {principal:,} | Rate: {annual_rate*100}% | Years: {years}')
    print(f'{"Year":<6} {"Amount":>12} {"Interest Earned":>16}')
    print('-' * 36)

    for year in range(1, years + 1):
        amount   = principal * (1 + annual_rate) ** year
        interest = amount - principal
        print(f'{year:<6} {amount:>12,.2f} {interest:>16,.2f}')

# Example: Rs 50 lakh invested at 8.5% for 10 years
compound_interest(5000000, 0.085, 10)

Principal: Rs 5,000,000 | Rate: 8.5% | Years: 10
Year         Amount  Interest Earned
------------------------------------
1      5,425,000.00       425,000.00
2      5,886,125.00       886,125.00
3      6,386,445.62     1,386,445.62
4      6,929,293.50     1,929,293.50
5      7,518,283.45     2,518,283.45
6      8,157,337.54     3,157,337.54
7      8,850,711.24     3,850,711.24
8      9,603,021.69     4,603,021.69
9      10,419,278.53     5,419,278.53
10     11,304,917.21     6,304,917.21


**Reading the table:** Notice how interest earned accelerates each year. By Year 10, the annual interest earned is far higher than Year 1 — even though the rate hasn't changed. This is the compounding effect in action.

## 2. EMI Calculator
An **EMI (Equated Monthly Instalment)** is the fixed monthly payment made to repay a loan over a set period. Every EMI payment covers two components:
- **Interest component** — cost of borrowing for that month
- **Principal component** — reduces the outstanding loan balance

In early months, most of the EMI goes toward interest. As the loan balance reduces, more goes toward principal.

**Formula:** EMI = P × r × (1+r)ⁿ / ((1+r)ⁿ - 1)
- P = loan principal
- r = monthly interest rate = annual rate ÷ 12
- n = number of monthly instalments

In [2]:
def loan_emi(principal, annual_rate, tenure_months):
    """
    Calculate EMI and total cost of a loan.
    principal      : loan amount (Rs)
    annual_rate    : annual interest rate as decimal
    tenure_months  : loan duration in months
    """
    r = annual_rate / 12       # convert annual rate to monthly
    n = tenure_months

    emi            = principal * r * (1 + r)**n / ((1 + r)**n - 1)
    total_paid     = emi * n
    total_interest = total_paid - principal

    print(f'Principal:       Rs {principal:,.2f}')
    print(f'Annual Rate:     {annual_rate*100}%')
    print(f'Tenure:          {tenure_months} months ({tenure_months//12} years)')
    print(f'Monthly EMI:     Rs {emi:,.2f}')
    print(f'Total Amount:    Rs {total_paid:,.2f}')
    print(f'Total Interest:  Rs {total_interest:,.2f}')

# Example: Rs 50 lakh home loan at 8.5% for 20 years
loan_emi(5000000, 0.085, 240)

Principal:       Rs 5,000,000.00
Annual Rate:     8.5%
Tenure:          240 months (20 years)
Monthly EMI:     Rs 43,391.16
Total Amount:    Rs 10,413,878.80
Total Interest:  Rs 5,413,878.80


**Key insight:** On a Rs 50 lakh loan at 8.5% for 20 years, you pay back over Rs 1 crore in total — more than double the borrowed amount. The extra Rs 54 lakhs is pure interest cost. This is why prepaying a home loan early saves a significant amount.

## 3. Moving Average
A **moving average** smooths out short-term price fluctuations to reveal the underlying trend direction. It is one of the most widely used indicators in technical analysis.

**How it works:** For a 3-day moving average, each output value is the average of the current day and the 2 preceding days. As each new day arrives, the window slides forward by one day.

**Common uses in finance:**
- **50-day SMA** — tracks medium-term trend
- **200-day SMA** — tracks long-term trend
- **Golden Cross** — when 50-day SMA crosses above 200-day SMA (bullish signal)
- **Death Cross** — when 50-day SMA crosses below 200-day SMA (bearish signal)

We will backtest a Golden Cross strategy on Nifty 50 in Week 4.

In [3]:
def moving_average(prices, window):
    """
    Compute simple moving average using a sliding window.
    prices : list of prices
    window : number of periods to average
    Returns a list of moving averages (shorter than input by window-1 values)
    """
    result = []
    for i in range(window - 1, len(prices)):
        window_prices = prices[i - window + 1 : i + 1]  # last 'window' prices
        avg = sum(window_prices) / window
        result.append(round(avg, 2))
    return result

# Test with 7 daily prices
prices = [100, 102, 98, 105, 110, 108, 112]
ma3 = moving_average(prices, window=3)

print(f'Prices:           {prices}')
print(f'3-day Moving Avg: {ma3}')
print(f'\nNote: First moving average starts from Day 3 (need 3 prices to compute first average)')
print(f'Input has {len(prices)} prices, output has {len(ma3)} averages')

Prices:           [100, 102, 98, 105, 110, 108, 112]
3-day Moving Avg: [100.0, 101.67, 104.33, 107.67, 110.0]

Note: First moving average starts from Day 3 (need 3 prices to compute first average)
Input has 7 prices, output has 5 averages


**Tracing through the calculation manually:**
- Day 3: avg(100, 102, 98) = 100.0
- Day 4: avg(102, 98, 105) = 101.67
- Day 5: avg(98, 105, 110) = 104.33
- Day 6: avg(105, 110, 108) = 107.67
- Day 7: avg(110, 108, 112) = 110.0

Each new day drops the oldest price and adds the newest — the window slides forward.

In [4]:
import numpy as np

# NumPy version using np.convolve — more efficient for large datasets
def moving_average_numpy(prices, window):
    weights = np.ones(window) / window
    return np.convolve(np.array(prices), weights, mode='valid')

ma3_numpy = moving_average_numpy(prices, 3)
print(f'NumPy version:    {ma3_numpy}')
print('Both versions produce identical results.')

NumPy version:    [100.         101.66666667 104.33333333 107.66666667 110.        ]
Both versions produce identical results.
